deepinv/training_utils/trainer.py

```diff
@@ -0,0 +1,714 @@
+        :ref:`See the libraries' evaluation metrics <loss>`.
+    :param float grad_clip: Gradient clipping value for the optimizer. If None, no gradient clipping is performed.
+    :param int ckp_interval: The model is saved every ``ckp_interval`` epochs.
+    :param int eval_interval: Number of epochs between each evaluation of the model on the evaluation set.
+    :param str save_path: Directory in which to save the trained model.
+    :param bool verbose: Output training progress information in the console.
+    :param bool plot_images: Plots reconstructions every ``ckp_interval`` epochs.
+    :param bool wandb_vis: Use Weights & Biases visualization, see https://wandb.ai/ for more details.
+    :param dict wandb_setup: Dictionary with the setup for wandb, see https://docs.wandb.ai/quickstart for more details.
+    :param bool plot_measurements: Plot the measurements y. default=True.
+    :param bool check_grad: Compute and print the gradient norm at each iteration.
+    :param str ckpt_pretrained: path of the pretrained checkpoint. If None, no pretrained checkpoint is loaded.
+    :param int freq_plot: Frequency of plotting images to wandb during train evaluation (at the end of each epoch).
+        If ``1``, plots at each epoch.
+    :param bool verbose_individual_losses: If ``True``, the value of individual losses are printed during training.
+        Otherwise, only the total loss is printed.
+    """
+
+    model: torch.nn.Module
+    physics: Union[Physics, List[Physics]]
+    optimizer: torch.optim.Optimizer
+    train_dataloader: torch.utils.data.DataLoader
+    epochs: int = 100
+    losses: Union[Loss, List[Loss]] = SupLoss()
+    eval_dataloader: torch.utils.data.DataLoader = None
+    scheduler: torch.optim.lr_scheduler.LRScheduler = None
+    metrics: Union[Loss, List[Loss]] = PSNR()
+    online_measurements: bool = False
+    grad_clip: float = None
+    ckp_interval: int = 1
+    eval_interval: int = 1
+    save_path: Union[str, Path] = "."
+    verbose: bool = True
+    plot_images: bool = False
+    plot_metrics: bool = False
+    wandb_vis: bool = False
+    wandb_setup: dict = field(default_factory=dict)
+    plot_measurements: bool = True
+    check_grad: bool = False
+    ckpt_pretrained: Union[str, None] = None
+    freq_plot: int = 1
+    verbose_individual_losses: bool = True
+    display_losses_eval: bool = False
+
+    def setup_train(self):
+        r"""
+        Set up the training process.
+
+        It initializes the wandb logging, the different metrics, the save path, the physics and dataloaders,
+        and the pretrained checkpoint if given.
+        """
+
+        if type(self.train_dataloader) is not list:
+            self.train_dataloader = [self.train_dataloader]
+
+        if self.eval_dataloader and type(self.eval_dataloader) is not list:
+            self.eval_dataloader = [self.eval_dataloader]
+
+        self.save_path = Path(self.save_path)
+
+        self.device = self.model.device
```

Comment:
I think we should keep device in the Trainer constructor rather than relying on the model to have a device - for example `deepinv.models.UNet` and `deepinv.models.SwinIR` both don't have `device` attributes

Reply:
I agree, this was a bad idea.